src/entity/Unit.ts

```diff
@@ -981,6 +984,23 @@ export function takeDamage(damageArgs: damageArgs, underworld: Underworld, predi
   // If taking damage (not healing) and health is 0 or less...
   if (amount > 0 && unit.health <= 0) {
     die(unit, underworld, prediction);
+    // Run onKill events for the sourceUnit of the lethal damage
+    const sourceUnit = damageArgs.sourceUnit;
+    if (sourceUnit) {
+      for (let eventName of sourceUnit.events) {
+        if (eventName) {
+          const fn = Events.onKillSource[eventName];
+          if (fn) {
+            fn(sourceUnit, unit, underworld, prediction);
+          }
+        }
+      }
+    }
+
+    // Special case: Bounty modifier persists on death and is removed AFTER onKill events run
+    if (unit.modifiers[bountyId]) {
+      removeModifier(unit, bountyId, underworld);
+    }
```

Comment:
Rather than have a special case here and need to add keepOnDeath again, let's just add `sourceUnit` to `die` and run the onKill event stuff inside of `die`.
What do you think?

Reply:
Oh, running the onKill events inside of die()  is _way_ better. I'll fix that.